pkg/receive/handler.go

```diff
@@ -1195,35 +1227,139 @@ func (p *peerGroup) close(addr string) error {
 		return fmt.Errorf("closing connection for %s", addr)
 	}
 
 	return nil
 }
 
-func (p *peerGroup) get(ctx context.Context, addr string) (storepb.WriteableStoreClient, error) {
+type peerWorkResponse struct {
+	er  endpointReplica
+	err error
+}
+
+func (p *peerWorker) RemoteWriteAsync(ctx context.Context, req *storepb.WriteRequest, er endpointReplica, seriesIDs []int, responseWriter chan writeResponse, cb func()) {
+	p.initWorkers()
+
+	w := peerWorkItem{
+		cc:         p.cc,
+		req:        req,
+		workResult: make(chan peerWorkResponse, 1),
+		ctx:        ctx,
+		er:         er,
+
+		sendTime: time.Now(),
+	}
+
+	p.work <- w
+	res := <-w.workResult
+
+	responseWriter <- newWriteResponse(seriesIDs, res.err, er)
+	cb()
+}
+
+func (p *peerGroup) get(ctx context.Context, addr string) (WriteableStoreAsyncClient, error) {
 	// use a RLock first to prevent blocking if we don't need to.
 	p.m.RLock()
 	c, ok := p.cache[addr]
 	p.m.RUnlock()
 	if ok {
-		return storepb.NewWriteableStoreClient(c), nil
+		return c, nil
 	}
 
 	p.m.Lock()
 	defer p.m.Unlock()
 	// Make sure that another caller hasn't created the connection since obtaining the write lock.
 	c, ok = p.cache[addr]
 	if ok {
-		return storepb.NewWriteableStoreClient(c), nil
+		return c, nil
 	}
 	conn, err := p.dialer(ctx, addr, p.dialOpts...)
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to dial peer")
 	}
 
-	p.cache[addr] = conn
-	return storepb.NewWriteableStoreClient(conn), nil
+	p.cache[addr] = newPeerWorker(conn, p.forwardDelay, p.asyncWorkerCount)
+	return p.cache[addr], nil
+}
+
+type peerWorker struct {
+	cc *grpc.ClientConn
+
+	work              chan peerWorkItem
+	turnOffGoroutines func()
+
+	initWorkersOnce  sync.Once
+	asyncWorkerCount uint
+	forwardDelay     prometheus.Histogram
+}
+
+func (pw *peerWorker) initWorkers() {
+	pw.initWorkersOnce.Do(func() {
+		work := make(chan peerWorkItem)
+		pw.work = work
+
+		ctx, cancel := context.WithCancel(context.Background())
+		pw.turnOffGoroutines = cancel
+
+		for i := 0; i < int(pw.asyncWorkerCount); i++ {
+			go func() {
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					case w := <-work:
+						pw.forwardDelay.Observe(time.Since(w.sendTime).Seconds())
+
+						tracing.DoInSpan(w.ctx, "receive_forward", func(ctx context.Context) {
+							_, err := storepb.NewWriteableStoreClient(pw.cc).RemoteWrite(ctx, w.req)
+							w.workResult <- peerWorkResponse{
+								er:  w.er,
+								err: errors.Wrapf(err, "forwarding request to endpoint %v", w.er.endpoint),
+							}
+							close(w.workResult)
+						})
+
+					}
+				}
+			}()
+		}
+
+	})
+}
+
+func newPeerWorker(cc *grpc.ClientConn, forwardDelay prometheus.Histogram, asyncWorkerCount uint) *peerWorker {
+	return &peerWorker{
+		cc:               cc,
+		asyncWorkerCount: asyncWorkerCount,
+		forwardDelay:     forwardDelay,
+	}
+}
+
+type peerWorkItem struct {
+	cc  *grpc.ClientConn
+	req *storepb.WriteRequest
+	ctx context.Context
```

Comment:
What's this context doing here? 🤔

Reply:
It's the context of the original request with a deadline + cancellation attached to it. Without this, the worker might spin forever. I'm not sure I understood the question

Reply:
I don't think it's a good idea to store the context like this in structs. Contexts aren't meant to be state. Maybe what we should be sending with each work item is the `ctx.Done()` channel?

Reply:
Okay, I'll make the change :+1:

Reply:
I was having a tough time to find where the context was being used. But I've found it.

Reply:
Actually I thought I had found it but I don't think I did. Can you help me understand this, @GiedriusS? I cannot clearly see where you use the work item's context.

Reply:
The confusion is **worker** context vs **work item** context.